### PR TITLE
Strings

### PR DESCRIPTION
**Pull request by hkgumbs**
One trickiness here is that the upstream compiler commits to an escape-based representation: https://github.com/hkgumbs/elm-ll/blob/master/compiler/src/Parse/Primitives.hs#L885-L895

This works upstream because they just wrap this string in JS single quotes and all is good. We get around this by manually unescaping the String as it is (but at least it's one pass, so that's nice).

We are also going to use a ByteString since that better maps to the Erlang binary type.

### CODE REVIEW

File: compiler/src/Generate/ErlangCore/Builder.hs [hunk 1]
```diff
@@ -64,6 +66,12 @@ fromFunction function =
 fromExpr :: Expr -> Builder
 fromExpr expression =
   case expression of
+    Int n ->
+      decimal n
+
+    Chr c ->
+      decimal (Char.ord c)
+
```
Comment (JeromeGoodrich) [inline compiler/src/Generate/ErlangCore/Builder.hs]: 👍

File: compiler/src/Generate/ErlangCore/Builder.hs [hunk 2]
```diff
@@ -83,18 +91,20 @@ fromExpr expression =
...
 
     List exprs ->
       "[" <> commaSep fromExpr exprs <> "]"
 
+    BitString str ->
+      let
+        bitChar c rest =
+          "#<" <> fromString (show c)
+          <> ">(8,1,'integer',['unsigned'|['big']])" : rest
+      in
```
Comment (JeromeGoodrich) [inline compiler/src/Generate/ErlangCore/Builder.hs]: Did you ever figure out what the '1' represents?

File: compiler/src/Generate/ErlangCore/Builder.hs [hunk 3]
```diff
@@ -83,18 +91,20 @@ fromExpr expression =
...
     Tuple exprs ->
       "{" <> commaSep fromExpr exprs <> "}"
 
     List exprs ->
       "[" <> commaSep fromExpr exprs <> "]"
 
+    BitString str ->
+      let
+        bitChar c rest =
+          "#<" <> fromString (show c)
```
Comment (JeromeGoodrich) [inline compiler/src/Generate/ErlangCore/Builder.hs]: Can we call this function introduced into the local scope something a bit more functiony? `toBitChar`maybe?

File: compiler/src/Generate/ErlangCore/String.hs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+
+bitString :: Text -> Core.Expr
+bitString =
+  Core.BitString . encodeUtf8 . unescape
+
+
+unescape :: Text -> Text
+unescape =
+  snd . Text.foldl unescapeHelp (No, Text.empty)
```
Comment (JeromeGoodrich) [inline compiler/src/Generate/ErlangCore/String.hs]: weird. I definitely thought you needed to import Tuple.

File: compiler/src/Generate/ErlangCore/String.hs [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+      (Yes, 'v') -> appendFinal '\v'
+
+      -- start escaping
+      (No, '\\') -> (Yes, textSoFar)
+
+      -- normal string character
+      _ -> appendFinal next
+
+
+data Escaping
```
Comment (JeromeGoodrich) [inline compiler/src/Generate/ErlangCore/String.hs]: Can we move this up for clarity? Not where I was expecting it.

File: compiler/src/Generate/ErlangCore/String.hs [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+  Core.BitString . encodeUtf8 . unescape
+
+
+unescape :: Text -> Text
+unescape =
+  snd . Text.foldl unescapeHelp (No, Text.empty)
+
+
+unescapeHelp :: (Escaping, Text) -> Char -> (Escaping, Text)
+unescapeHelp (escaping, textSoFar) next =
```
Comment (JeromeGoodrich) [inline compiler/src/Generate/ErlangCore/String.hs]: 👏 I bet this was tricky to figure out. Good job!

File: compiler/tests/test-files/good/Literals/QuotesAndComments.elm [hunk 7]
```diff
@@ -7,4 +7,4 @@ singleQuote = '\''
 
 
 {- this is an {- embeded comment -} -}
```
Comment (JeromeGoodrich) [inline compiler/tests/test-files/good/Literals/QuotesAndComments.elm]: not sure if it matters, but `embedded` is spelled wrong here.

File: compiler/tests/test-files/good/Literals/Unicode.elm [hunk 8]
```diff
@@ -0,0 +1,12 @@
+
+singleTab = '\v'
+
+multipleTabs = "\t\v"
+
+newline = '\n'
+
+unicodeCharacter = '🔥'
```
Comment (JeromeGoodrich) [inline compiler/tests/test-files/good/Literals/Unicode.elm]: 🔥

File: compiler/src/Generate/ErlangCore/Builder.hs (hunk 2 above)
Comment (hkgumbs) [inline compiler/src/Generate/ErlangCore/Builder.hs]: It's just called `unit` [here](https://github.com/erlang/otp/blob/master/lib/compiler/src/core_pp.erl#L528-L533)... So no

Comment (JeromeGoodrich) [inline compiler/src/Generate/ErlangCore/Builder.hs]: lol yeah...

**Pull request merged**